\VC\include\xutility(364): error C2039: 'iterator_category' : is not a member of 'dealii::IndexSet::ElementIterator'

When compiling the file \dealii\source\multigrid\mg_tools.cc I got the error. I compared the current code with the code of 8.4.1 version to find the reason. xutility is used by VS 2013 in inner implementation of vector and set headers. the expression that causes the error is in line 1192 in mg_tools.cc:
boundary_indices[i].insert(boundary_indexset[i].begin(), boundary_indexset[i].end());
in 8.4.1 you had:
boundary_indices[i].add_indices(boundary_indexset[i].begin(), boundary_indexset[i].end());
It means that you used your own implementation of insert function.
The problem is that now you use your own iterators in standard container without deriving them from standard iterators. Realization of iterators is platform dependent so you have to declare your class
ElementIterator to inherit standard iterator which is used in vector or to continue using your own implementation add_indices. I can not see other options. I attach a complete log of this error to the post.
[ElementIterator_issue.txt](https://github.com/dealii/dealii/files/384825/ElementIterator_issue.txt)

This is the error (copied from the attached file for easier reading):
```
2>  mg_tools.cc
2>C:\Program Files (x86)\Microsoft Visual Studio 11.0\VC\include\xutility(364): error C2039: 'iterator_category' : is not a member of 'dealii::IndexSet::Elemen
tIterator'
2>          H:\DealII\source\dealii\include\deal.II/base/index_set.h(583) : see declaration of 'dealii::IndexSet::ElementIterator'
2>          C:\Program Files (x86)\Microsoft Visual Studio 11.0\VC\include\xutility(649) : see reference to class template instantiation 'std::iterator_traits<
_Iter>' being compiled
2>          with
2>          [
2>              _Iter=dealii::IndexSet::ElementIterator
2>          ]
2>          C:\Program Files (x86)\Microsoft Visual Studio 11.0\VC\include\set(177) : see reference to function template instantiation 'void std::_Debug_range<
_Iter>(_InIt,_InIt,std::_Dbfile_t,std::_Dbline_t)' being compiled
2>          with
2>          [
2>              _Iter=dealii::IndexSet::ElementIterator,
2>              _InIt=dealii::IndexSet::ElementIterator
2>          ]
2>          H:\DealII\source\dealii\source\multigrid\mg_tools.cc(1193) : see reference to function template instantiation 'void std::set<_Kty>::insert<dealii::
IndexSet::ElementIterator>(_Iter,_Iter)' being compiled
2>          with
2>          [
2>              _Kty=unsigned int,
2>              _Iter=dealii::IndexSet::ElementIterator
2>          ]
2>          H:\DealII\source\dealii\source\multigrid\mg_tools.cc(1193) : see reference to function template instantiation 'void std::set<_Kty>::insert<dealii::
IndexSet::ElementIterator>(_Iter,_Iter)' being compiled
2>          with
2>          [
2>              _Kty=unsigned int,
2>              _Iter=dealii::IndexSet::ElementIterator
2>          ]
2>          H:\DealII\source\dealii\build\source\multigrid\mg_tools.inst(1761) : see reference to function template instantiation 'void dealii::MGTools::make_b
oundary_list<2,2>(const dealii::DoFHandler<dim> &,const std::map<_Kty,_Ty> &,std::vector<std::set<unsigned int>> &,const dealii::ComponentMask &)' being compil
ed
2>          with
2>          [
2>              dim=2,
2>              _Kty=dealii::types::boundary_id,
2>              _Ty=const dealii::Function<2,double> *
2>          ]
2>C:\Program Files (x86)\Microsoft Visual Studio 11.0\VC\include\xutility(364): error C2146: syntax error : missing ';' before identifier 'iterator_category'
2>C:\Program Files (x86)\Microsoft Visual Studio 11.0\VC\include\xutility(364): error C4430: missing type specifier - int assumed. Note: C++ does not support default-int
```

--- a/include/deal.II/base/index_set.h
+++ b/include/deal.II/base/index_set.h
@@ -22,6 +22,8 @@
 #include <boost/serialization/vector.hpp>
 #include <vector>
 #include <algorithm>
+#include <iterator>
+
 
 #ifdef DEAL_II_WITH_TRILINOS
 #  include <Epetra_Map.h>
@@ -496,7 +498,7 @@ public:
    * Class that represents an iterator pointing to a contiguous interval
    * $[a,b[$ as returned by IndexSet::begin_interval().
    */
-  class IntervalIterator
+  class IntervalIterator : public std::iterator<std::forward_iterator_tag,IntervalAccessor>
   {
   public:
     /**
@@ -579,14 +581,16 @@ public:
    * Class that represents an iterator pointing to a single element in the
    * IndexSet as returned by IndexSet::begin().
    */
-  class ElementIterator
+  class ElementIterator : public std::iterator<std::forward_iterator_tag,size_type>
   {
   public:
     /**
      * Construct an iterator pointing to the global index @p index in the
      * interval @p range_idx
      */
-    ElementIterator(const IndexSet *idxset, const size_type range_idx, const size_type index);
+    ElementIterator(const IndexSet *idxset,
+                    const size_type range_idx,
+                    const size_type index);
 
     /**
      * Construct an iterator pointing to the end of the IndexSet.
@@ -1129,7 +1133,7 @@ inline
 IndexSet::ElementIterator
 IndexSet::ElementIterator::operator++ (int)
 {
-  IndexSet::ElementIterator it = *this;
+  const IndexSet::ElementIterator it = *this;
   advance();
   return it;
 }

